WalletWasabi/Blockchain/TransactionOutputs/SmartCoin.cs

```diff
@@ -323,9 +313,31 @@ public TxoRef GetTxoRef()
 
 		public bool Equals(SmartCoin other) => this == other;
 
-		public override int GetHashCode() => (TransactionId, Index).GetHashCode();
+		public override int GetHashCode() => HashCode;
 
-		public static bool operator ==(SmartCoin x, SmartCoin y) => y?.TransactionId == x?.TransactionId && y?.Index == x?.Index;
+		public static bool operator ==(SmartCoin x, SmartCoin y)
+		{
+			if (ReferenceEquals(x, y))
+			{
+				return true;
+			}
+			else if (x is null)
+			{
+				return false;
+			}
+			else
+			{
+				if (y is null)
+				{
+					return false;
+				}
+				else
+				{
+					var hashEquals = x.HashCode == y.HashCode;
+					return hashEquals && y?.TransactionId == x?.TransactionId && y?.Index == x?.Index;
+				}
+			}
```

Comment:
I suggest to write it like this:
```
if (ReferenceEquals(x, y))
{
	return true;
}
else if (x is null)
{
	return false;
}
else if (y is null)
{
	return false;
}
else
{
	var hashEquals = x.HashCode == y.HashCode;
	return hashEquals && y?.TransactionId == x?.TransactionId && y?.Index == x?.Index;
}
```
OR
```
if (ReferenceEquals(x, y))
{
	return true;
}
else if (x is null || y is null)
{
	return false;
}
else
{
	var hashEquals = x.HashCode == y.HashCode;
	return hashEquals && y?.TransactionId == x?.TransactionId && y?.Index == x?.Index;
}
```

Reply:
Thanks, introducing referenceequals removed a few code paths and I didn't notice the optimization.